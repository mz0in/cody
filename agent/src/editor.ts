import type * as vscode from 'vscode'
import { URI } from 'vscode-uri'

import { isCodyIgnoredFile } from '@sourcegraph/cody-shared/src/chat/context-filter'
import {
    ActiveTextEditor,
    ActiveTextEditorDiagnostic,
    ActiveTextEditorSelection,
    ActiveTextEditorViewControllers,
    ActiveTextEditorVisibleContent,
    Editor,
} from '@sourcegraph/cody-shared/src/editor'

import { TextDocumentWithUri } from '../../vscode/src/jsonrpc/TextDocumentWithUri'

import { Agent } from './agent'
import { DocumentOffsets } from './offsets'

export class AgentEditor implements Editor {
    public controllers?: ActiveTextEditorViewControllers | undefined

    constructor(private agent: Agent) {}

    public didReceiveFixupText(): Promise<void> {
        throw new Error('Method not implemented.')
    }

    /** @deprecated Use {@link AgentEditor.getWorkspaceRootUri} instead. */
    public getWorkspaceRootPath(): string | null {
        const uri = this.getWorkspaceRootUri()
        return uri?.scheme === 'file' ? uri.fsPath : null
    }

    public getWorkspaceRootUri(): vscode.Uri | null {
        return this.agent.workspace.workspaceRootUri ?? null
    }

    private activeDocument(): TextDocumentWithUri | undefined {
        if (this.agent.workspace.activeDocumentFilePath === null) {
            return undefined
        }
        if (isCodyIgnoredFile(URI.file(this.agent.workspace.activeDocumentFilePath.fsPath))) {
            return undefined
        }
        return this.agent.workspace.getDocument(this.agent.workspace.activeDocumentFilePath)?.underlying
    }

    public getActiveTextEditor(): ActiveTextEditor | null {
        const document = this.activeDocument()
        if (document === undefined) {
            return null
        }
        return {
            filePath: document.uri.fsPath,
            fileUri: document.uri,
            selectionRange: document.selection,
            content: document.content || '',
        }
    }

    public async getTextEditorContentForFile(uri: URI): Promise<string | undefined> {
        if (!uri) {
            return Promise.resolve(undefined)
        }

        const doc = this.agent.workspace.getDocumentFromUriString(uri.toString())
        return Promise.resolve(doc?.getText())
    }

    public getActiveTextEditorSelection(): ActiveTextEditorSelection | null {
        const document = this.activeDocument()
        if (document?.content === undefined || document.selection === undefined) {
            return null
        }
        const offsets = new DocumentOffsets(document.underlying)
        if (!document.selection) {
            return {
                fileName: document.uri.fsPath,
                fileUri: document.uri,
                selectionRange: document.selection,
                precedingText: document.content ?? '',
                selectedText: '',
                followingText: '',
            }
        }
        const from = offsets.offset(document.selection.start)
        const to = offsets.offset(document.selection.end)
        return {
            fileName: document.uri.fsPath,
            fileUri: document.uri,
            selectionRange: document.selection,
            precedingText: document.content.slice(0, from),
            selectedText: document.content.slice(from, to),
            followingText: document.content.slice(to, document.content.length),
        }
    }

    public getActiveTextEditorSelectionOrEntireFile(): ActiveTextEditorSelection | null {
        const document = this.activeDocument()
        if (document !== undefined && document.selection === undefined) {
            return {
                fileName: document.uri.fsPath,
                fileUri: document.uri,
                precedingText: '',
                selectedText: document.content || '',
                followingText: '',
            }
        }
        return this.getActiveTextEditorSelection()
    }

    public getActiveTextEditorSmartSelection(): Promise<ActiveTextEditorSelection | null> {
        throw new Error('Method not implemented.')
    }

    public getActiveTextEditorSelectionOrVisibleContent(): ActiveTextEditorSelection | null {
        throw new Error('Method not implemented.')
    }

    public getActiveTextEditorDiagnosticsForRange(): ActiveTextEditorDiagnostic[] | null {
        throw new Error('Method not implemented.')
    }

    public getActiveTextEditorVisibleContent(): ActiveTextEditorVisibleContent | null {
        const document = this.activeDocument()
        if (document === undefined) {
            return null
        }
        return {
            content: document.content || '',
            fileName: document.uri.fsPath,
            fileUri: document.uri,
        }
    }

    public replaceSelection(): Promise<void> {
        throw new Error('Not implemented')
    }

    public showQuickPick(): Promise<string | undefined> {
        throw new Error('Not implemented')
    }

    public showWarningMessage(): Promise<void> {
        throw new Error('Not implemented')
    }

    public showInputBox(): Promise<string | undefined> {
        throw new Error('Not implemented')
    }
}
