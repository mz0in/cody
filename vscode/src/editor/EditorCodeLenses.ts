import * as vscode from 'vscode'

import { getEditor } from './active-editor'

interface EditorCodeLens {
    name: string
    selection: vscode.Selection
}

/**
 * Adds Code lenses for triggering Recipes Menu and inline Chat (when enabled)
 * on top of all the functions in active documents
 */
export class EditorCodeLenses implements vscode.CodeLensProvider {
    private isEnabled = false

    private _disposables: vscode.Disposable[] = []
    private _onDidChangeCodeLenses: vscode.EventEmitter<void> = new vscode.EventEmitter<void>()
    public readonly onDidChangeCodeLenses: vscode.Event<void> = this._onDidChangeCodeLenses.event
    constructor() {
        this.provideCodeLenses = this.provideCodeLenses.bind(this)
        this.updateConfig()
        vscode.workspace.onDidChangeConfiguration(e => {
            if (e.affectsConfiguration('cody')) {
                this.updateConfig()
            }
        })
    }

    /**
     * init
     */
    private init(): void {
        if (!this.isEnabled) {
            return
        }
        this._disposables.push(vscode.languages.registerCodeLensProvider({ scheme: 'file' }, this))
        this._disposables.push(
            vscode.commands.registerCommand('cody.editor.codelens.click', async lens => {
                const clickedLens = lens as EditorCodeLens
                await this.onCodeLensClick(clickedLens)
            })
        )
        // on change events for toggling
        this._disposables.push(
            vscode.window.onDidChangeVisibleTextEditors(() => this.fire()),
            vscode.window.onDidChangeActiveTextEditor(() => this.fire())
        )
    }

    /**
     * Update the configurations
     */
    private updateConfig(): void {
        const config = vscode.workspace.getConfiguration('cody')
        this.isEnabled = config.get('commandCodeLenses') as boolean

        if (this.isEnabled && !this._disposables.length) {
            this.init()
        }
        this.fire()
    }

    /**
     * Handle the code lens click event
     */
    private async onCodeLensClick(lens: EditorCodeLens): Promise<void> {
        // Update selection in active editor to the selection of the clicked code lens
        const activeEditor = getEditor().active
        if (activeEditor) {
            activeEditor.selection = lens.selection
        }
        await vscode.commands.executeCommand(lens.name, 'codeLens')
    }
    /**
     * Gets the code lenses for the specified document.
     */
    public async provideCodeLenses(
        document: vscode.TextDocument,
        token: vscode.CancellationToken
    ): Promise<vscode.CodeLens[]> {
        if (!this.isEnabled) {
            return []
        }
        token.onCancellationRequested(() => [])
        const editor = getEditor().active
        if (!editor || editor.document !== document || document.languageId === 'json') {
            return []
        }
        // Generate code lenses for the document.
        const codeLenses = []
        const codeLensesMap = new Map<string, vscode.Range>()

        // Get a list of symbols from the document, filter out symbols that are not functions / classes / methods
        const allSymbols = await vscode.commands.executeCommand<vscode.SymbolInformation[]>(
            'vscode.executeDocumentSymbolProvider',
            document.uri
        )
        const symbols = allSymbols?.filter(
            symbol =>
                symbol.kind === vscode.SymbolKind.Function ||
                symbol.kind === vscode.SymbolKind.Class ||
                symbol.kind === vscode.SymbolKind.Method ||
                symbol.kind === vscode.SymbolKind.Constructor
        )

        // Add code lenses for each symbol
        if (symbols) {
            for (const symbol of symbols) {
                const range = symbol.location.range
                const selection = new vscode.Selection(range.start, range.end)
                codeLenses.push(
                    new vscode.CodeLens(range, {
                        ...editorCodeLenses.cody,
                        arguments: [{ name: 'cody.action.commands.menu', selection }],
                    })
                )
                codeLensesMap.set(symbol.location.range.start.line.toString(), range)
            }
        }

        return codeLenses
    }

    /**
     * Fire an event to notify VS Code that the code lenses have changed.
     */
    public fire(): void {
        if (!this.isEnabled) {
            this.dispose()
            return
        }
        this._onDidChangeCodeLenses.fire()
    }

    /**
     * Dispose the disposables
     */
    public dispose(): void {
        if (this._disposables.length) {
            for (const disposable of this._disposables) {
                disposable.dispose()
            }
            this._disposables = []
        }
        this._onDidChangeCodeLenses.fire()
    }
}

const editorCodeLenses = {
    cody: { title: '$(cody-logo) Cody', command: 'cody.editor.codelens.click', tooltip: 'Open command menu' },
    inline: { title: 'Inline Chat', command: 'cody.editor.codelens.click', tooltip: 'Ask Cody inline' },
}
