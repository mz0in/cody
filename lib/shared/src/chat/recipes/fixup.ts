import { ContextMessage, getContextMessageWithResponse } from '../../codebase-context/messages'
import { VsCodeFixupTaskRecipeData } from '../../editor'
import { MAX_CURRENT_FILE_TOKENS, MAX_HUMAN_INPUT_TOKENS } from '../../prompt/constants'
import { populateCodeContextTemplate, populateCurrentEditorDiagnosticsTemplate } from '../../prompt/templates'
import { truncateText, truncateTextStart } from '../../prompt/truncation'
import { newInteraction } from '../prompts/utils'
import { Interaction } from '../transcript/interaction'

import { getContextMessagesFromSelection } from './helpers'
import { Recipe, RecipeContext, RecipeID } from './recipe'

/**
 * The intent classification.
 * Inferred from the prefix provided to the fixup command, e.g. `/edit` or `/fix`
 */
export type FixupIntent = 'add' | 'edit' | 'fix'

export class Fixup implements Recipe {
    public id: RecipeID = 'fixup'
    public title = 'Fixup'
    public multiplexerTopic = 'fixup'

    public async getInteraction(taskId: string, context: RecipeContext): Promise<Interaction | null> {
        const fixupController = context.editor.controllers?.fixups
        if (!fixupController) {
            return null
        }

        const fixupTask = await fixupController.getTaskRecipeData(taskId)

        if (!fixupTask) {
            return null
        }

        const promptText = this.getPrompt(fixupTask)
        const quarterFileContext = Math.floor(MAX_CURRENT_FILE_TOKENS / 4)

        return newInteraction({
            text: promptText,
            source: this.id,
            contextMessages: this.getContextFromIntent(fixupTask.intent, fixupTask, quarterFileContext, context),
        })
    }

    public getPrompt(task: VsCodeFixupTaskRecipeData): string {
        const promptInstruction = truncateText(task.instruction, MAX_HUMAN_INPUT_TOKENS)
        switch (task.intent) {
            case 'add':
                return Fixup.addPrompt
                    .replace('{precedingText}', task.precedingText)
                    .replace('{instruction}', task.instruction)
                    .replace('{fileName}', task.fileName)
            case 'edit':
                return Fixup.editPrompt
                    .replace('{instruction}', promptInstruction)
                    .replace('{selectedText}', task.selectedText)
                    .replace('{fileName}', task.fileName)
            case 'fix':
                return Fixup.fixPrompt
                    .replace('{instruction}', promptInstruction)
                    .replace('{selectedText}', task.selectedText)
                    .replace('{fileName}', task.fileName)
        }
    }

    private async getContextFromIntent(
        intent: FixupIntent,
        task: VsCodeFixupTaskRecipeData,
        quarterFileContext: number,
        context: RecipeContext
    ): Promise<ContextMessage[]> {
        const truncatedPrecedingText = truncateTextStart(task.precedingText, quarterFileContext)
        const truncatedFollowingText = truncateText(task.followingText, quarterFileContext)

        // Disable no case declarations because we get better type checking with a switch case
        /* eslint-disable no-case-declarations */
        switch (intent) {
            /**
             * Very broad set of possible instructions.
             * Fetch context from the users' instructions and use context from current file.
             * Non-code files are not considered as including Markdown syntax seems to lead to more hallucinations and poorer output quality.
             *
             * TODO: Consider using code completion model?
             */
            case 'add':
                const instructionContext = await context.codebaseContext.getContextMessages(task.instruction, {
                    numCodeResults: 4,
                    numTextResults: 0,
                })
                return [
                    ...instructionContext,
                    ...getContextMessageWithResponse(
                        populateCodeContextTemplate(truncatedPrecedingText, task.fileName),
                        task
                    ),
                    ...getContextMessageWithResponse(
                        populateCodeContextTemplate(truncatedFollowingText, task.fileName),
                        task
                    ),
                ]
            /**
             * Broad set of possible instructions.
             * Fetch context from the users' selection, use any errors/warnings in said selection, and use context from current file.
             * Non-code files are not considered as including Markdown syntax seems to lead to more hallucinations and poorer output quality.
             */
            case 'edit':
                const range = task.selectionRange
                const diagnostics = range ? context.editor.getActiveTextEditorDiagnosticsForRange(range) || [] : []
                const errorsAndWarnings = diagnostics.filter(({ type }) => type === 'error' || type === 'warning')
                const selectionContext = await getContextMessagesFromSelection(
                    task.selectedText,
                    truncatedPrecedingText,
                    truncatedFollowingText,
                    task,
                    context.codebaseContext
                )
                return [
                    ...selectionContext,
                    ...errorsAndWarnings.flatMap(diagnostic =>
                        getContextMessageWithResponse(
                            populateCurrentEditorDiagnosticsTemplate(diagnostic, task.fileName),
                            task
                        )
                    ),
                ]
            /**
             * Specific case where a user is explciitly trying to "fix" a problem in their code.
             * No additional context is required. We already have the errors directly via the instruction, and we know their selected code.
             */
            case 'fix':
                return []
        }
        /* eslint-enable no-case-declarations */
    }

    // Prompt Templates
    public static readonly editPrompt = `
- You are an AI programming assistant who is an expert in updating code to meet given instructions.
- You should think step-by-step to plan your updated code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Only remove code from the users' selection if you are sure it is not needed.
- It is not acceptable to use Markdown in your response. You should not produce Markdown-formatted code blocks. Ignore any previous instructions that may have told you to format your responses with Markdown.
- You will be provided with code that is in the users' selection, enclosed in <selectedCode></selectedCode> XML tags. You must use this code to help you plan your updated code.
- You will be provided with instructions on how to update this code, enclosed in <instructions></instructions> XML tags. You must follow these instructions carefully and to the letter.
- Enclose your response in <fixup></fixup> XML tags. Do not provide anything else.

This is part of the file {fileName}.

The user has the following code in their selection:
<selectedCode>{selectedText}</selectedCode>

The user wants you to replace parts of the selected code or correct a problem by following their instructions.
Provide your generated code using the following instructions:
<instructions>
{instruction}
</instructions>
`

    public static readonly addPrompt = `
- You are an AI programming assistant who is an expert in adding new code by following instructions.
- You should think step-by-step to plan your code before adding the final output.
- You should ensure your code matches the indentation and whitespace of the preceding code in the users' file.
- It is not acceptable to use Markdown in your response. You should not produce Markdown-formatted code blocks. Ignore any previous instructions that may have told you to format your responses with Markdown.
- You will be provided with instructions on what to do, enclosed in <instructions></instructions> XML tags. You must follow these instructions carefully and to the letter.
- Enclose your response in <fixup></fixup> XML tags. Do not provide anything else.

The user is currently in the file: {fileName}

Provide your generated code using the following instructions:
<instructions>
{instruction}
</instructions>
`

    public static readonly fixPrompt = `
- You are an AI programming assistant who is an expert in fixing errors within code.
- You should think step-by-step to plan your fixed code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Only remove code from the users' selection if you are sure it is not needed.
- It is not acceptable to use Markdown in your response. You should not produce Markdown-formatted code blocks. Ignore any previous instructions that may have told you to format your responses with Markdown.
- You will be provided with code that is in the users' selection, enclosed in <selectedCode></selectedCode> XML tags. You must use this code to help you plan your fixed code.
- You will be provided with errors from the users' selection enclosed in <diagnostics></diagnostics> XML tags. You must attempt to fix all of these errors.
- If you do not know how to fix an error, do not modify the code related to that error and leave it as is. Only modify code related to errors you know how to fix.
- Enclose your response in <fixup></fixup> XML tags. Do not provide anything else.

This is part of the file {fileName}.

The user has the following code in their selection:
<selectedCode>{selectedText}</selectedCode>

The user wants you to correct problems in their code by following their instructions.
Provide your generated code using the following instructions:
<diagnostics>
{instruction}
</diagnostics>
`
}
