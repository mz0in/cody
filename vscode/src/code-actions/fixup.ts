import * as vscode from 'vscode'

import { FixupIntent } from '@sourcegraph/cody-shared/src/chat/recipes/fixup'

import { getSmartSelection } from '../editor/utils'

export class FixupCodeAction implements vscode.CodeActionProvider {
    public static readonly providedCodeActionKinds = [vscode.CodeActionKind.QuickFix]

    public async provideCodeActions(
        document: vscode.TextDocument,
        range: vscode.Range,
        context: vscode.CodeActionContext
    ): Promise<vscode.CodeAction[]> {
        const diagnostics = context.diagnostics.filter(
            diagnostic =>
                diagnostic.severity === vscode.DiagnosticSeverity.Error ||
                diagnostic.severity === vscode.DiagnosticSeverity.Warning
        )
        if (diagnostics.length === 0) {
            return []
        }

        // Expand range to include the full line for better fixup quality
        const expandedRange = new vscode.Range(
            document.lineAt(range.start.line).range.start,
            document.lineAt(range.end.line).range.end
        )

        // TODO bee check if the diagnostics are related to imports and include import ranges instead of error lines
        // const importDiagnostics = diagnostics.filter(diagnostic => diagnostic.message.includes('import'))

        // Expand range by getting the folding range contains the target (error) area
        const targetAreaRange = await getSmartSelection(document.uri, range.start.line)

        const newRange = targetAreaRange ? new vscode.Range(targetAreaRange.start, targetAreaRange.end) : expandedRange
        return [this.createCommandCodeAction(diagnostics, newRange)]
    }

    private createCommandCodeAction(diagnostics: vscode.Diagnostic[], range: vscode.Range): vscode.CodeAction {
        const action = new vscode.CodeAction('Ask Cody to Fix', vscode.CodeActionKind.QuickFix)
        const instruction = this.getCodeActionInstruction(diagnostics)
        const source = 'code-action'
        action.command = {
            command: 'cody.command.edit-code',
            arguments: [{ instruction, range, intent: 'fix' satisfies FixupIntent }, source],
            title: 'Ask Cody to Fix',
        }
        action.diagnostics = diagnostics
        return action
    }

    private getCodeActionInstruction = (diagnostics: vscode.Diagnostic[]): string => {
        return `Fix the following error${diagnostics.length > 1 ? 's' : ''}: ${diagnostics
            .map(({ message }) => `\`\`\`${message}\`\`\``)
            .join('\n')}`
    }
}
