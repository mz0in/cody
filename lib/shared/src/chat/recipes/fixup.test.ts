import { describe, expect, it } from 'vitest'

import { VsCodeFixupTaskRecipeData } from '../../editor'

import { Fixup } from './fixup'

describe('Fixup', () => {
    const fixupTask: Omit<VsCodeFixupTaskRecipeData, 'intent'> = {
        fileName: 'src/file/index.ts',
        followingText: "const text = 'Hello, world!'\n",
        selectedText: 'return text',
        precedingText: '\n}',
        instruction: 'Console log text',
        selectionRange: { start: { line: 2, character: 0 }, end: { line: 2, character: 11 } },
    }

    it('builds prompt correctly for edits', () => {
        const fixup = new Fixup()
        expect(fixup.getPrompt({ ...fixupTask, intent: 'edit' })).toMatchSnapshot()
    })

    it('builds prompt correctly for adding', () => {
        const fixup = new Fixup()
        expect(fixup.getPrompt({ ...fixupTask, intent: 'add' })).toMatchSnapshot()
    })

    it('builds prompt correctly for fixing', () => {
        const fixup = new Fixup()
        expect(fixup.getPrompt({ ...fixupTask, intent: 'fix' })).toMatchSnapshot()
    })
})
