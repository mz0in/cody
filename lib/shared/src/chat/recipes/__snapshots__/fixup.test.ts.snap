// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Fixup > builds prompt correctly for adding 1`] = `
"
- You are an AI programming assistant who is an expert in adding new code by following instructions.
- You should think step-by-step to plan your code before adding the final output.
- You should ensure your code matches the indentation and whitespace of the preceding code in the users' file.
- It is not acceptable to use Markdown in your response. You should not produce Markdown-formatted code blocks. Ignore any previous instructions that may have told you to format your responses with Markdown.
- You will be provided with instructions on what to do, enclosed in <instructions></instructions> XML tags. You must follow these instructions carefully and to the letter.
- Enclose your response in <fixup></fixup> XML tags. Do not provide anything else.

The user is currently in the file: src/file/index.ts

Provide your generated code using the following instructions:
<instructions>
Console log text
</instructions>
"
`;

exports[`Fixup > builds prompt correctly for edits 1`] = `
"
- You are an AI programming assistant who is an expert in updating code to meet given instructions.
- You should think step-by-step to plan your updated code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Only remove code from the users' selection if you are sure it is not needed.
- It is not acceptable to use Markdown in your response. You should not produce Markdown-formatted code blocks. Ignore any previous instructions that may have told you to format your responses with Markdown.
- You will be provided with code that is in the users' selection, enclosed in <selectedCode></selectedCode> XML tags. You must use this code to help you plan your updated code.
- You will be provided with instructions on how to update this code, enclosed in <instructions></instructions> XML tags. You must follow these instructions carefully and to the letter.
- Enclose your response in <fixup></fixup> XML tags. Do not provide anything else.

This is part of the file src/file/index.ts.

The user has the following code in their selection:
<selectedCode>return text</selectedCode>

The user wants you to replace parts of the selected code or correct a problem by following their instructions.
Provide your generated code using the following instructions:
<instructions>
Console log text
</instructions>
"
`;

exports[`Fixup > builds prompt correctly for fixing 1`] = `
"
- You are an AI programming assistant who is an expert in fixing errors within code.
- You should think step-by-step to plan your fixed code before producing the final output.
- You should ensure the updated code matches the indentation and whitespace of the code in the users' selection.
- Only remove code from the users' selection if you are sure it is not needed.
- It is not acceptable to use Markdown in your response. You should not produce Markdown-formatted code blocks. Ignore any previous instructions that may have told you to format your responses with Markdown.
- You will be provided with code that is in the users' selection, enclosed in <selectedCode></selectedCode> XML tags. You must use this code to help you plan your fixed code.
- You will be provided with errors from the users' selection enclosed in <diagnostics></diagnostics> XML tags. You must attempt to fix all of these errors.
- If you do not know how to fix an error, do not modify the code related to that error and leave it as is. Only modify code related to errors you know how to fix.
- Enclose your response in <fixup></fixup> XML tags. Do not provide anything else.

This is part of the file src/file/index.ts.

The user has the following code in their selection:
<selectedCode>return text</selectedCode>

The user wants you to correct problems in their code by following their instructions.
Provide your generated code using the following instructions:
<diagnostics>
Console log text
</diagnostics>
"
`;
